### PR TITLE
rgw: add support bucket replication between zonegroups

### PR DESCRIPTION
**Pull request by clwluvw**
Currently, only bucket replication is supported inside a zonegroup between zones. This doesn't respect the bucket location-constraints. With supporting bucket replication based on buckets between zonegroups, an AWS model can be achieved by respecting the location constraints and only replicating buckets based on users' requests by PutBucketReplication API (https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketReplication.html).

The current sync pipes are effectively designed to halt logging of all operations, irrespective of the sync policy applied per bucket. By extending the implementation to support multiple zonegroups, zones will be able to pull data from other zones across different zonegroups.

Fixes: https://tracker.ceph.com/issues/66649

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by clwluvw** [comment]
If I have understand correctly, currently the only overhead is every zone will process all logged buckets from the source zone until they but just by the name and they will stop at https://github.com/ceph/ceph/blob/cbbddfd41a357d0a727b75f320a4b152ce2d98b8/src/rgw/driver/rados/rgw_data_sync.cc#L5392-L5395
I'm not sure how expensive this could be.

**Comment by clwluvw** [comment]
I think I misunderstood the purpose of `RGWDataChangesLog`. It seems that it only logs the bi log marker, so the ideas mentioned earlier are not suitable for this implementation. I think the best approach would be to enable `rgw_data_change` to store the destination zone, and when other zones are polling for the logs, they can filter the entries by the `dest_zone`. This filtering operation should be inexpensive for the source zone.

**Comment by smanjara** [comment]
thanks @clwluvw. does replication between zonegroups work with the changes from df1867f0a16a6ad67ea4ef7f7d4ec1721bbe8fa8 alone? if it does, we might want to do some testing to see how expensive the polling would be under a reasonable load and a reasonable number of zones before we optimize for its efficiency. 
two zonegroup testing was disabled in our multisite testing suite due to failing test cases. now we run a subset of the whole suite and most of them should pass. I have enabled two zonegroup yaml here: https://github.com/ceph/ceph/pull/60172

**Comment by clwluvw** [comment]
Hi @smanjara. Thanks for the intrest. I'll break my commits as below:

- df1867f0a16a6ad67ea4ef7f7d4ec1721bbe8fa8 will enable having zonegroups in the conn map. but the replication would still be blocked as `log_data` is set to false when we have only one zone in the zonegroup (so this alone needs a modification to set `log_data: true` manually in the zonegroup config.
- 9fb96a9f2bc000d95c4479a658808279ca948435 - this will eliminate the need for manually activation by checking whether the bucket has any active pipes rather than checking that boolean. but this will add the load for loading bucket sync pipe per any operation that before was checking by only a const bool. (but I checked and there is a sycn pipe cache already in-place so hopefully this shouldn't introduce any significant perf drop)
- f980c03d27e5ffdcac60e04477ea028bd6f8f54e - this is the same as above but I forgot to fix it for `renew_entries()` so you can assume these two are squashed.
- 8f87e0a932433171b9ac25933cd2e79ed7eae208 - this will conder per object than bucket to whether to log an object or not. it should have the same cost as per bucket checking I belive as the check per object should be just a loop by rules checking prefix and obj tags (sometimes it loads object tags but those are for apis that probably response time should not matter in difference x ms).

So to summarise the first three commit should let you replicate without any manual work but the last one will optimize to only log if the rule has any prefix/tag filter.

**Comment by clwluvw** [comment]
> how expensive the polling would be under a reasonable load and a reasonable number of zones before we optimize for its efficiency.

With the current implementation, as I understand from (https://github.com/ceph/ceph/pull/59911#issuecomment-2372215270), other zones in different zonegroups will just list the datalog and ignore entries when they are not interested in a particular bucket. This would likely affect the `sync status` more, though it's unclear how much of the destination zone's resources would be consumed, which might lead to issues. Your load test could help determine this need.

There’s one incompatibility with AWS BucketReplication: the priority field in the [ReplicationRule](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ReplicationRule.html#AmazonS3-Type-ReplicationRule-Priority). AWS only replicates to one bucket per object, and if there are collisions with the rules defined on the source bucket, it selects the highest priority. Currently, RGW does not respect this priority, which might confuse users who rely on it.

However, with the proposed approach outlined here (https://github.com/ceph/ceph/pull/59911#issuecomment-2395629449), I believe we can address this issue while also reducing log processing by destination zones. In the logging phase, we can define where the object should be replicated. If I'm not mistaken, this might also replace the need for the `zones_trace` concept currently implemented.

**Comment by smanjara** [comment]
@clwluvw sorry it took me so long to get back.  the commit https://github.com/ceph/ceph/commit/df1867f0a16a6ad67ea4ef7f7d4ec1721bbe8fa8 establishes connection objects for zones across zonegroups when zonegroup policy is set. this alone along with bucket location constraint fixes https://github.com/ceph/ceph/pull/59305 and https://github.com/ceph/ceph/pull/59960 should be sufficient to setup replication between two zonegroups.

 along with zonegroup policy set to `allowed`, we need to `enable` bucket sync policy to either sync within zonegroup or between zones across zonegroups. although I am not sure how sync between the zones on non-master zonegroup will behave and what the sync status will look like. this will need some testing and investigation.

documenting configuration example here for reference. 

at the zonegroup level:

```
$radosgw-admin sync group create  --group-id=group1 --status=allowed
$radosgw-admin sync group flow create  --group-id=group1 --flow-id=flow-mirror --flow-type=symmetrical --zones=zg1-1,zg1-2,zg2-3,zg2-4
$radosgw-admin sync group pipe create --group-id=group1 --pipe-id=pipe1 --source-zones='*' --source-bucket='*' --dest-zones='*' --dest-bucket='*'
$radosgw-admin period update --commit
```
enable bucket1 replication between zg1-2 and zg1-2 belonging to zonegroup zg1. 

```
$radosgw-admin sync group create --bucket=bucket1 --group-id=bucket1-default --status=enabled
$radosgw-admin sync group flow create --bucket=bucket1 --group-id=bucket1-default --flow-id=bucket1-flow --flow-type=symmetrical --zones="zg1-1, zg1-2"
$radosgw-admin sync group pipe create --bucket=bucket1 --group-id=bucket1-default --pipe-id=bucket1-pipe --source-zones='*' --dest-zones='*'
```

enable bucket3 replication to sync between the zonegroups zg and zg2 involving all zones:

```
$radosgw-admin sync group create --bucket=bucket3 --group-id=bucket3-default --status=enabled
$radosgw-admin sync group flow create --bucket=bucket3 --group-id=bucket3-default --flow-id=bucket3-flow --flow-type=symmetrical --zones="zg1-1, zg1-2, zg2-3, zg2-4"
$radosgw-admin sync group pipe create --bucket=bucket3 --group-id=bucket3-default --pipe-id=bucket3-pipe --source-zones='*' --dest-zones='*'
```

please note that setting zonegroup sync policy `allowed` forces you to set sync policy on each bucket individually.
we can further reduce the number of zones participating probably by adding only one of the zones from zonegroup zg2 in the bucket sync flow, if you don't need redundancy within the zonegroup and you only care about data locality. 

there is a commit in https://github.com/ceph/ceph/pull/60018 that sets multiple zonegroups for you to test with.

the other commits deal with changing the way we log data, that I am not very comfortable with. the most common multisite configuration is the one where we sync to/from all zones within a zonegroup. adding conditional checks for sync pipes or for specific objects may not work and adds an overhead for configurations that does not care about sync policies.

**Comment by clwluvw** [comment]
Thank you for your time on this @smanjara.

> the commit [df1867f](https://github.com/ceph/ceph/commit/df1867f0a16a6ad67ea4ef7f7d4ec1721bbe8fa8) establishes connection objects for zones across zonegroups when zonegroup policy is set. this alone along with bucket location constraint fixes #59305 and #59960 should be sufficient to setup replication between two zonegroups.

I'm not sure if this would be enough. Basically as long as [`log_data`](https://github.com/ceph/ceph/blob/20af41d716086a13009d3404acdee6c7c1444ae0/src/rgw/rgw_zone_types.h#L321) is not true on the zonegroup config, RGW should not log anything: https://github.com/ceph/ceph/blob/20af41d716086a13009d3404acdee6c7c1444ae0/src/rgw/driver/rados/rgw_rados.cc#L7319-L7325
Currently `log_data` will be enabled when we have more than one zone in within the RGW's zonegroup https://github.com/ceph/ceph/blob/20af41d716086a13009d3404acdee6c7c1444ae0/src/rgw/driver/rados/rgw_zone.cc#L234 So I guess in your test case it accidentally worked because you had always more than one zone per zonegroup(?) and if you have a zonegroup with only one zone it shouldn't have `log_data` set to true. that's why the other two commits are needed I believe.

> along with zonegroup policy set to `allowed`, we need to `enable` bucket sync policy to either sync within zonegroup or between zones across zonegroups. although I am not sure how sync between the zones on non-master zonegroup will behave and what the sync status will look like. this will need some testing and investigation.

Right, from the concept I was thinking of the logging and polling mechanism independently and replaying them based on the filters and bucket's zonegroup availability. So I guess it should not matter whether the source or dest is master.

> please note that setting zonegroup sync policy `allowed` forces you to set sync policy on each bucket individually. we can further reduce the number of zones participating probably by adding only one of the zones from zonegroup zg2 in the bucket sync flow, if you don't need redundancy within the zonegroup and you only care about data locality.

Right, currently AWS is also designed in a way that you can only replicate an object to a single bucket (zonegroup) and not concurrently to more. but still, that is not enough here, as you said before, the entry processing would be done by all zones no matter what the policy says, but only one will do that actual replication in terms of data.

> the other commits deal with changing the way we log data, that I am not very comfortable with. the most common multisite configuration is the one where we sync to/from all zones within a zonegroup. adding conditional checks for sync pipes or for specific objects may not work and adds an overhead for configurations that does not care about sync policies.

Currently, the same check per object based on the sync policies available is done here: https://github.com/ceph/ceph/blob/20af41d716086a13009d3404acdee6c7c1444ae0/src/rgw/driver/rados/rgw_datalog.cc#L662-L664
The only difference that I'm making now is to limit the filter to the object's properties (`prefix` and `tags`) and only log if the pipe has any interest in that object particularly (which can be an `enabled` pipe on the bucket itself or on the zonegroup level). So, the load of processing the sync pipe is already being done (the pipe loading has a cache already though), but I'm just adding a minor check based on the filters, which I guess shouldn't cause a significant performance drop. Do you see it in another way?

**Comment by smanjara** [comment]
> Thank you for your time on this @smanjara.
> 
> > the commit [df1867f](https://github.com/ceph/ceph/commit/df1867f0a16a6ad67ea4ef7f7d4ec1721bbe8fa8) establishes connection objects for zones across zonegroups when zonegroup policy is set. this alone along with bucket location constraint fixes #59305 and #59960 should be sufficient to setup replication between two zonegroups.
> 
> I'm not sure if this would be enough. Basically as long as [`log_data`](https://github.com/ceph/ceph/blob/20af41d716086a13009d3404acdee6c7c1444ae0/src/rgw/rgw_zone_types.h#L321) is not true on the zone config, RGW should not log anything:
> 
> https://github.com/ceph/ceph/blob/20af41d716086a13009d3404acdee6c7c1444ae0/src/rgw/driver/rados/rgw_rados.cc#L7319-L7325
> 
> 
> Currently `log_data` will be enabled when we have more than one zone in within the RGW's zonegroup
> https://github.com/ceph/ceph/blob/20af41d716086a13009d3404acdee6c7c1444ae0/src/rgw/driver/rados/rgw_zone.cc#L234
> 
> So I guess in your test case it accidentally worked because you had always more than one zone per zonegroup(?) and if you have a zonegroup with only one zone it shouldn't have `log_data` set to true. that's why the other two commits are needed I believe.

yeah I am looking at it as an extension of multisite where we will simply add new zones from other zonegroups seamlessly while 'log_data' is true on all zones. I have added your pr as a topic of discussion here: https://pad.ceph.com/p/rgw-weekly. hopefull you can make it.
cc @yehudasa

**Comment by clwluvw** [comment]
There is also another challenge regarding replicating an already replicated object. Currently, we log and replicate even when an object is being replicated again, which increases the load and adds complexity in managing `zones_trace`.

The main issue is that when an object is replicated via `RGWBucketSyncSingleEntryCR()`, we log it again because another active sync process matches this operation. If that sync points back to the original zone, it becomes redundant. I haven't found a clean way (even with the `log_zonegroup` implementation, which is insufficient since it might involve another zone) to stop logging when it's a circular replication. For now, the zones_trace logic simply ignores it during polling, but that still results in an unnecessary load.

AWS S3 doesn't support this type of replication, as described here: https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication-what-is-isnot-replicated.html
> Objects in the source bucket that are replicas that were created by another replication rule. For example, suppose you configure replication where bucket A is the source and bucket B is the destination. Now suppose that you add another replication configuration where bucket B is the source and bucket C is the destination. In this case, objects in bucket B that are replicas of objects in bucket A are not replicated to bucket C.

For performance and efficiency, could we consider dropping this replication? Or at least maybe we introduce an option in the configuration to disable this functionality.

**Comment by clwluvw** [comment]
So now I guess I could reflect my idea described here https://github.com/ceph/ceph/pull/59911#issuecomment-2395629449 but with a different name called `log_zonegroup`. Basically, all datalog and bi logs will contain a `log_zonegroup` which shows which zonegroup has the interest, and on `rgw_rest_log` API calls the source zone will filter based on `rgwx-zonegroup` provided by the dest zone.

To wrap up, basically, we perhaps can say with these changes, bi-logging and datalogchange-logging will only happen if there is any `enabled` sync pipe matched to the object/bucket (this can be a pipe for a specific bucket or the zone overall) and on the polling phase, the destination zone will only receive and process the related logs to itself and not the other zonegroups logs.

**Comment by clwluvw** [comment]
@cbodley - regarding the scaling issues on the polling phase, just wanted to mention that the `log_zonegroup` idea is also integrated into the `rgw_data_change` struct. So if your concern was around listing the `bucket-instance` then it should be already addressed but if it's literally around the "HTTP calls" particularly, then right for more than 20 zonegroups it might cause scaling issues with a low number of RGWs.

**Comment by clwluvw** [comment]
Hi @cbodley, @smanjara - I have added a description about the proposal here in the PR's description (I changed the `log_zonegroup` idea to `log_zones` based on the description due to trimming issues on the old approach).

The tests I've added are passing on my local with 3 zonegroups. This should be ready for the first round of review. Thanks.

**Comment by clwluvw** [comment]
Apologies for the noise, everyone! I made a mistake with the rebase (hopefully, the drafting already unsubscribed you).

Sorry to mention you again, @cbodley @smanjara — I’ve refactored the commits (dropped the old design) in hopes it’ll make it easier for review.

**Comment by clwluvw** [comment]
> @clwluvw i'm curious about how we handle 'data full sync' between zones in different zonegroups

I guess for cross-zonegroup replication `rgw_data_sync_info::StateBuildingFullSyncMaps` could be meaningless. i.e. from our discussion here (https://github.com/ceph/ceph/pull/61003#discussion_r1889146590), we verify the policies getting added by making sure the dest bucket exists (i.e. the dest bucket at that point is a bucket that already resists on one of the existing _zonegorups_). If I understand correctly `rgw_data_sync_info::StateBuildingFullSyncMaps` only happens on new zone deployment, therefore if it's about a new _zonegroup_ then there should be no policies pointing to that zonegroup where it hasn't existed at all, and if it's about a new zone in a zonegroup then due to the symmetric replication concept within a zonegroup, we should create sync index only for the buckets that the zonegroup matches our zonegroup.

@cbodley So perhaps we should change that part to limit the buckets to the RGW's zonegroup, right?

**Comment by cbodley** [comment]
> @cbodley So perhaps we should change that part to limit the buckets to the RGW's zonegroup, right?

that sounds right 👍

File: src/cls/rgw/cls_rgw_types.cc [hunk 1]
```diff
@@ -238,6 +238,7 @@ void rgw_bucket_dir_entry::generate_test_instances(list<rgw_bucket_dir_entry*>&
     e->exists = true;
     e->meta = *m;
     e->tag = "tag";
+    e->log_zones = { rgw_zone_id("1588bb2c-439a-4b75-91ef-f0b31d02563b"), rgw_zone_id("1f9654c2-3a66-4407-b07c-0f6727c9df17") };
```
Comment (anthonyeleven) [inline src/cls/rgw/cls_rgw_types.cc:241]: Is it intentional to hardcode these UUIDs?

File: src/common/options/rgw.yaml.in [hunk 2]
```diff
@@ -4274,3 +4274,17 @@ options:
   services:
   - rgw
   with_legacy: true
+- name: rgw_data_sync_allow_chain_replication
+  type: bool
+  level: advanced
+  default: true
+  desc: This option controls whether replication of already replicated objects (chain replication)
```
Comment (anthonyeleven) [inline src/common/options/rgw.yaml.in:4281]: suggest

already-replicated

File: src/cls/rgw/cls_rgw_types.cc (hunk 1 above)
Comment (clwluvw) [reply src/cls/rgw/cls_rgw_types.cc:241]: This is a function that generates test data.
